Complete S5 MovieQuiz

Evo-Eclipse opened the pull request:
Modularization of Models and Services:
- Added new separated files for model structures: `AlertModel`, `GameResult`, `QuizQuestion`, `QuizResultsViewModel`, and `QuizStepViewModel` in the `Models` directory
- Crafted `QuestionFactory` and `QuestionFactoryDelegate` to manage quiz question generation and delegation
- Created `ResultAlertPresenter` to handle alert presentation logic

View Controller Refactoring:
- Refactored `MovieQuizViewController` to remove hardcoded models and constants, replacing them with the newly created models and services
- Integrated `QuestionFactory` to dynamically fetch quiz questions and handle the quiz flow
- Updated the display logic to use `ResultAlertPresenter` for showing quiz results

Evo-Eclipse committed:
Complete S5 MovieQuiz

Andreynnt commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -144,71 +101,63 @@ final class MovieQuizViewController: UIViewController {
-        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
+        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) { [weak self] in
+            guard let self = self else { return }
             self.showNextQuestionOrResults()
```
Тут необязательно разворачивать self-a, `self?.showNextQuestionOrResults()` будет эквивалентно. guard делают обычно, когда есть последовательность/несколько обращений к self-у, и чтобы все они выполнились - чтобы не было ситуации:
```
self?.doSmth1()
self?.doSmth1()
// затем self стал nil
self?.doSmth3() - уже не выполнится
```
Это некритично, можно так и так, просто будет чуть короче

Andreynnt commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -144,71 +101,63 @@ final class MovieQuizViewController: UIViewController {
     private func showNextQuestionOrResults() {
-        if currentQuestionIndex == questions.count - 1 {
-            updateGameStats()
+        if currentQuestionIndex == questionsAmount - 1 {
```
Лучше использовать в подобных конструкциях guard вместо if-a. Так будет меньше вложенность кода, соответсвенно будет проще читать

Andreynnt commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -0,0 +1,60 @@
+class QuestionFactory: QuestionFactoryProtocol {
```
Стоит добавить модификатор final, чтобы вызовы методов работали быстрее. 
Но это уже со звездочкой, хотя на собеседованиях спрашивают. Можно почитать про диспачеризацию тут https://habr.com/ru/articles/673400/

Evo-Eclipse committed:
Review Fixes

- Replaced unnecessary `guard` with optional chaining in `DispatchQueue` usage
- Refactored `if` to `guard` in `showNextQuestionOrResults` for reduced nesting
- Added `final` modifier to `QuestionFactory` for improved performance
